From 529de7ea28b1a3541cb182425a65543d79c0d820 Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Wed, 25 Feb 2026 10:17:10 +0000
Subject: [PATCH] Allow system_app to access 8849 projector and laser sysfs

Add SELinux policy to allow TrebleApp (system_app) to read and write
the FPGA SPI sysfs nodes at /sys/bus/spi/drivers/fpga_spi/ which
control the Unihertz 8849 (Tank) built-in projector hardware, and to
read the NDS03 laser rangefinder sysfs node at /sys/nds03_ctrl/.

This covers: projector_pitch_setting (keystone), projector_rgb_current
(brightness), projector_fan_en (fan speed), motor_turn_mode_step
(focus), focus_auto_calibration, projector_sys_status, projector_temp,
dlpc3421_awb_cali, fpga_h3_scan_hori, fpga_h3_scan_vert, nds03_read.

Without this policy, writes from system_app context are denied:
  avc: denied { write } for comm="sh" name="projector_pitch_setting"
  scontext=u:r:system_app:s0 tcontext=u:object_r:sysfs:s0
  avc: denied { read } for name="nds03_read"
  scontext=u:r:system_app:s0 tcontext=u:object_r:sysfs:s0
---
 sepolicy/genfs_contexts | 2 ++
 sepolicy/treble_app.te  | 9 +++++++++
 2 files changed, 11 insertions(+)

diff --git a/sepolicy/genfs_contexts b/sepolicy/genfs_contexts
index 56e8950..05ad092 100644
--- a/sepolicy/genfs_contexts
+++ b/sepolicy/genfs_contexts
@@ -1 +1,3 @@
  genfscon sysfs /board_properties u:object_r:sysfs_board_properties:s0
 +genfscon sysfs /bus/spi/drivers/fpga_spi u:object_r:sysfs_fpga_spi:s0
 +genfscon sysfs /nds03_ctrl u:object_r:sysfs_nds03:s0
diff --git a/sepolicy/treble_app.te b/sepolicy/treble_app.te
index 4bed5f1..9ea193c 100644
--- a/sepolicy/treble_app.te
+++ b/sepolicy/treble_app.te
@@ -39,3 +39,12 @@ binder_call(system_app, hal_gesturewake_default);
 
 type hal_aidl_agolddaemon_default, domain;
 binder_call(system_app, hal_aidl_agolddaemon_default);
+
+# Unihertz 8849 (Tank) projector FPGA sysfs access.
+# The projector is controlled via /sys/bus/spi/drivers/fpga_spi/ nodes
+# for pitch correction, brightness, fan speed, focus motor, etc.
+type sysfs_fpga_spi, fs_type, sysfs_type;
+allow system_app sysfs_fpga_spi:dir r_dir_perms;
+allow system_app sysfs_fpga_spi:file rw_file_perms;
+
+# Unihertz 8849 (Tank) NDS03 laser rangefinder sysfs access.
+type sysfs_nds03, fs_type, sysfs_type;
+allow system_app sysfs_nds03:file r_file_perms;
-- 
2.53.0

